### PR TITLE
feat:`GetNews`

### PR DESCRIPTION
**Pull request by CommanderStorm**
This PR implements the news api `GetNews`

### CODE REVIEW

File: server/api/tumdev/campus_backend.proto [hunk 1]
```diff
@@ -376,6 +383,27 @@ message Room {
...
+}
+
+message GetNewsReply {
+  repeated NewsItem news = 1;
+}
+
+message GetNewsRequest {
+  // the last id of the news item received. 0 to get all news items
+  int32 last_news_id = 1;
+  // filter by news source id. 0 or empty to get all news items
```
Comment (kordianbruck) [inline server/api/tumdev/campus_backend.proto:403]: nit: "empty" is not an option. Unset might be.

File: server/backend/cron/news.go [hunk 2]
```diff
@@ -120,7 +122,8 @@ func (c *CronService) parseNewsFeed(source model.NewsSource) error {
 				Src:         source.Source,
 				Link:        item.Link,
 				Image:       enclosureUrl,
-				File:        fileId,
+				FilesID:     null.Int{NullInt64: sql.NullInt64{Int64: int64(file.File), Valid: true}},
```
Comment (kordianbruck) [inline server/backend/cron/news.go:125]: This seems a bit manual, just to create a nullable int. There must be a shorthand for this. See https://pkg.go.dev/github.com/guregu/null#IntFrom

File: server/backend/cron/news.go [hunk 3]
```diff
@@ -9,6 +9,8 @@ import (
 	"strings"
 	"time"
 
+	"gorm.io/gorm/clause"
+
```
Comment (kordianbruck) [inline server/backend/cron/news.go:13]: Generally you just separate external from std imports and proto imports. You can just keep this in one block together with the github imports.

File: server/backend/cron/news.go [hunk 4]
```diff
@@ -137,39 +140,23 @@ func (c *CronService) parseNewsFeed(source model.NewsSource) error {
...
-		return null.Int{}, err
+	if err := c.db.Clauses(clause.OnConflict{
+		Columns:   []clause.Column{{Name: "path"}},
+		DoNothing: true,
+	}).Create(&file).Error; err != nil {
+		log.WithError(err).WithField("targetFileName", targetFileName).Error("Couldn't query database for file")
+		return nil, err
 	}
-	// creating this int is annoying but i'm too afraid to use real ORM in the model
-	return null.Int{NullInt64: sql.NullInt64{Int64: int64(file.File), Valid: true}}, nil
```
Comment (kordianbruck) [inline server/backend/cron/news.go:172]: How is the file.File (the ID) actually populated here?

Comment (kordianbruck) [reply server/backend/cron/news.go:172]: You seemingly access that field outside of this function, but its never populated here.

File: server/backend/news.go [hunk 5]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"fmt"
 
+	"google.golang.org/protobuf/types/known/timestamppb"
```
Comment (kordianbruck) [inline server/backend/news.go:7]: Again, three blocks: std, external and proto imports

File: server/backend/news.go [hunk 6]
```diff
@@ -34,3 +36,39 @@ func (s *CampusServer) GetNewsSources(ctx context.Context, _ *emptypb.Empty) (ne
...
+
+func (s *CampusServer) GetNews(ctx context.Context, req *pb.GetNewsRequest) (newsSources *pb.GetNewsReply, err error) {
+	if err = s.checkDevice(ctx); err != nil {
+		return
+	}
+
+	var sources []model.News
+	// preloading files as the news images can be null
+	// I could not get outer joins to work => we are currently doing unnecessary db calls
+	tx := s.db.Preload("Files")
```
Comment (kordianbruck) [inline server/backend/news.go:48]: nit: well this table could be really big eventually, so idk?

File: server/backend/news.go [hunk 7]
```diff
@@ -34,3 +36,39 @@ func (s *CampusServer) GetNewsSources(ctx context.Context, _ *emptypb.Empty) (ne
 	}
 	return &pb.NewsSourceReply{Sources: resp}, nil
 }
+
+func (s *CampusServer) GetNews(ctx context.Context, req *pb.GetNewsRequest) (newsSources *pb.GetNewsReply, err error) {
+	if err = s.checkDevice(ctx); err != nil {
+		return
+	}
+
+	var sources []model.News
```
Comment (kordianbruck) [inline server/backend/news.go:45]: These aren't "news sources" but actual news entries, no? Seems the name for the variable is a bit off.

File: server/model/files.go [hunk 8]
```diff
@@ -18,7 +18,7 @@ var (
 // Files struct is a row record of the files table in the tca database
 type Files struct {
 	File       int32          `gorm:"primary_key;AUTO_INCREMENT;column:file;type:int;" json:"file"`
-	Name       string         `gorm:"column:name;type:text;size:16777215;" json:"name"`
+	Name       string         `gorm:"index:idx_files_name,unique;column:name;type:text;size:16777215;" json:"name"`
```
Comment (kordianbruck) [inline server/model/files.go:21]: Just the name of the file probably isn't actually unique. The combination of the name and the path might be.

Comment (kordianbruck) [reply server/model/files.go:21]: (like there was a reason for the complexity in saveimage)

File: server/api/tumdev/campus_backend.proto (hunk 1 above)
Comment (CommanderStorm) [reply server/api/tumdev/campus_backend.proto:403]: Given that it is unclear how other programing languages implment unset values, lets just document 0

```suggestion
  // filter by news source id. 0 to get all news items
```

File: server/backend/cron/news.go (hunk 2 above)
Comment (CommanderStorm) [reply server/backend/cron/news.go:125]: Good catch!
I assumed this was just one of the go-quirks
⇒ #229 (and rebasing tomorrow 😉)

File: server/backend/cron/news.go (hunk 3 above)
Comment (CommanderStorm) [reply server/backend/cron/news.go:13]: The import grouping/formatting is done by [`goimports`](https://pkg.go.dev/golang.org/x/tools/cmd/goimports) => if the current import formatting is bad, I would like to change it in another PR

File: server/backend/news.go (hunk 6 above)
Comment (CommanderStorm) [reply server/backend/news.go:48]: Why is this always so difficult?
image

You are correct => Raw query it is

File: server/backend/cron/news.go (hunk 4 above)
Comment (CommanderStorm) [reply server/backend/cron/news.go]: I reverted this part.
I don't know how but I am pretty sure that I checked if the filename is indeed an unique characteristic (apparetnly not properly)

File: server/model/files.go (hunk 8 above)
Comment (CommanderStorm) [reply server/model/files.go]: I reverted this part.
I don't know how but I am pretty sure that I checked if the filename is indeed an unique characteristic (apparetnly not properly)

File: server/backend/cron/news.go [hunk 9]
```diff
@@ -139,20 +137,28 @@ func (c *CronService) parseNewsFeed(source model.NewsSource) error {
...
 }
 
-// saveImage Saves an image to the database, so it can be downloaded by another cronjob
+// saveImage Saves an image to the database so it can be downloaded by another cronjob and returns its id
 func (c *CronService) saveImage(url string) (*model.Files, error) {
 	targetFileName := fmt.Sprintf("%x.jpg", md5.Sum([]byte(url)))
 	file := model.Files{
+		Name: targetFileName,
+	}
+	if err := c.db.Model(model.Files{}).First(&file).Error; err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
```
Comment (kordianbruck) [inline server/backend/cron/news.go:146]:
```suggestion
	if err := c.db.First(&file).Error; err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
```

File: server/backend/cron/news.go [hunk 10]
```diff
@@ -139,20 +137,28 @@ func (c *CronService) parseNewsFeed(source model.NewsSource) error {
 	return nil
 }
 
-// saveImage Saves an image to the database, so it can be downloaded by another cronjob
+// saveImage Saves an image to the database so it can be downloaded by another cronjob and returns its id
 func (c *CronService) saveImage(url string) (*model.Files, error) {
 	targetFileName := fmt.Sprintf("%x.jpg", md5.Sum([]byte(url)))
 	file := model.Files{
+		Name: targetFileName,
```
Comment (kordianbruck) [inline server/backend/cron/news.go:144]:
```suggestion
		Name: targetFileName, // path intentionally omitted
```

**Review by kordianbruck — APPROVED** [review]

**Pull request merged**